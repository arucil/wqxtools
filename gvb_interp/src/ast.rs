use crate::parser::ParseResult;
use smallvec::{Array, SmallVec};
use std::cmp::Ordering;
use std::fmt::{self, Debug, Formatter};
use std::num::NonZeroUsize;
use std::ops::{Deref, DerefMut};

pub mod expr;
pub mod label;
pub mod line;
pub mod node;
pub mod stmt;
pub mod token;

pub use self::expr::*;
pub use self::label::*;
pub use self::line::*;
pub use self::node::*;
pub use self::stmt::*;
pub use self::token::*;

#[derive(Clone)]
pub struct Program {
  pub lines: Vec<ParseResult<ProgramLine>>,
}

pub struct NonEmptyVec<T: Array>(pub SmallVec<T>);

#[derive(Clone, PartialEq, Eq)]
pub struct Range {
  pub start: CharIndex,
  pub end: CharIndex,
}

#[derive(Clone, Eq, Ord)]
pub struct CharIndex {
  pub utf8: u32,
  pub utf16: u32,
}

#[derive(Clone, PartialEq, Eq)]
pub struct CharOffset {
  pub utf8: i32,
  pub utf16: i32,
}

impl<T: Array> PartialEq for NonEmptyVec<T>
where
  T::Item: PartialEq,
{
  fn eq(&self, other: &Self) -> bool {
    self.0.eq(&other.0)
  }
}

impl<T: Array> Eq for NonEmptyVec<T> where T::Item: Eq {}

impl<T: Array> NonEmptyVec<T> {
  pub fn len(&self) -> NonZeroUsize {
    unsafe { NonZeroUsize::new_unchecked(self.0.len()) }
  }
}

impl<T: Array> Clone for NonEmptyVec<T>
where
  T::Item: Clone,
{
  fn clone(&self) -> Self {
    Self(self.0.clone())
  }
}

impl<T: Array> Debug for NonEmptyVec<T>
where
  T::Item: Debug,
{
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    self.0.fmt(f)
  }
}

#[cfg(test)]
impl Program {
  pub fn to_string(&self, text: &str) -> String {
    let mut buf = String::new();
    let mut offset = 0;
    for line in &self.lines {
      buf += &line.to_string(&text[offset..offset + line.content.source_len]);
      buf += "==================================\n";
      offset += line.content.source_len;
    }
    buf
  }
}

impl<T: Array> NonEmptyVec<T> {
  pub fn new() -> Self {
    Self(SmallVec::new())
  }
}

impl<T: Array> Deref for NonEmptyVec<T> {
  type Target = SmallVec<T>;
  fn deref(&self) -> &Self::Target {
    &self.0
  }
}

impl<T: Array> DerefMut for NonEmptyVec<T> {
  fn deref_mut(&mut self) -> &mut Self::Target {
    &mut self.0
  }
}

impl Range {
  pub fn new(start: CharIndex, end: CharIndex) -> Self {
    assert!(start <= end);
    Self { start, end }
  }

  pub fn new_ascii(start: u32, end: u32) -> Self {
    assert!(start <= end);
    Self {
      start: CharIndex::new_ascii(start),
      end: CharIndex::new_ascii(end),
    }
  }

  pub fn empty(start: CharIndex) -> Self {
    Self {
      start: start.clone(),
      end: start,
    }
  }

  pub fn start() -> Self {
    Self {
      start: CharIndex::start(),
      end: CharIndex::start(),
    }
  }

  pub fn empty_ascii(start: u32) -> Self {
    Self {
      start: CharIndex::new_ascii(start),
      end: CharIndex::new_ascii(start),
    }
  }

  pub fn is_empty(&self) -> bool {
    self.start == self.end
  }

  pub fn utf8_len(&self) -> u32 {
    self.end.utf8 - self.start.utf8
  }

  pub fn utf16_len(&self) -> u32 {
    self.end.utf16 - self.start.utf16
  }

  pub fn offset(&self, offset: CharOffset) -> Self {
    Self {
      start: self.start.offset(offset),
      end: self.end.offset(offset),
    }
  }

  pub fn utf8_range(&self) -> std::ops::Range<usize> {
    (self.start.utf8 as usize)..(self.end.utf8 as usize)
  }
}

impl Debug for Range {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    f.pad(&format!("{:?}..{:?}", self.start, self.end))
  }
}

impl CharIndex {
  pub fn new_ascii(i: u32) -> Self {
    Self { utf8: i, utf16: i }
  }

  pub fn start() -> Self {
    Self { utf8: 0, utf16: 0 }
  }

  pub fn offset(&self, offset: CharOffset) -> Self {
    Self {
      utf8: self.utf8 + offset.utf8 as u32,
      utf16: self.utf16 + offset.utf16 as u32,
    }
  }

  pub fn end_of<S>(s: S) -> Self
  where
    S: AsRef<str>,
  {
    let s = s.as_ref();
    Self {
      utf8: s.len() as u32,
      utf16: s.encode_utf16().count() as u32,
    }
  }
}

impl Debug for CharIndex {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    f.pad(&format!("({}|{})", self.utf8, self.utf16))
  }
}

impl PartialEq for CharIndex {
  fn eq(&self, other: &Self) -> bool {
    self.utf8 == other.utf8
  }
}

impl PartialOrd for CharIndex {
  fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
    self.utf8.partial_cmp(&other.utf8)
  }
}

impl From<char> for CharOffset {
  fn from(value: char) -> Self {
    Self {
      utf8: value.len_utf8() as i32,
      utf16: value.len_utf16() as i32,
    }
  }
}

impl CharOffset {
  pub fn new_ascii(n: i32) -> Self {
    Self { utf8: n, utf16: n }
  }
}
