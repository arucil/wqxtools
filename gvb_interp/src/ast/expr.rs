use id_arena::Arena;

use super::{ExprId, NonEmptyVec, Range};
use num_derive::FromPrimitive;
use std::fmt::{self, Debug, Formatter, Write};
use std::str::FromStr;

#[derive(Debug, Clone)]
pub struct Expr {
  pub kind: ExprKind,
  pub range: Range,
}

#[derive(Debug, Clone)]
pub enum ExprKind {
  Ident,
  StringLit,
  NumberLit,
  SysFuncCall {
    func: (Range, SysFuncKind),
    args: NonEmptyVec<[ExprId; 1]>,
  },
  UserFuncCall {
    /// ident
    func: Option<Range>,
    arg: ExprId,
  },
  Binary {
    lhs: ExprId,
    op: (Range, BinaryOpKind),
    rhs: ExprId,
  },
  Unary {
    op: (Range, UnaryOpKind),
    arg: ExprId,
  },
  Index {
    name: Option<Range>,
    indices: NonEmptyVec<[ExprId; 1]>,
  },
  Inkey,
  Error,
}

#[derive(Clone, Copy, PartialEq, Eq, FromPrimitive)]
pub enum SysFuncKind {
  Abs,
  Asc,
  Atn,
  Chr,
  Cos,
  Cvi,
  Cvs,
  Eof,
  Exp,
  Int,
  Left,
  Len,
  Lof,
  Log,
  Mid,
  Mki,
  Mks,
  Peek,
  Pos,
  Right,
  Rnd,
  Sgn,
  Sin,
  Sqr,
  Str,
  Tan,
  Val,
}

#[derive(Clone, Copy, PartialEq, Eq)]
pub enum BinaryOpKind {
  Eq,
  Ne,
  Gt,
  Lt,
  Ge,
  Le,
  Add,
  Sub,
  Mul,
  Div,
  Pow,
  And,
  Or,
}

#[derive(Clone, Copy, PartialEq, Eq)]
pub enum UnaryOpKind {
  Not,
  Neg,
  Pos,
}

impl Expr {
  pub fn new(kind: ExprKind, range: Range) -> Self {
    Self { kind, range }
  }
}

impl FromStr for SysFuncKind {
  type Err = ();
  fn from_str(s: &str) -> Result<Self, ()> {
    match s {
      "abs" => Ok(Self::Abs),
      "asc" => Ok(Self::Asc),
      "atn" => Ok(Self::Atn),
      "chr$" => Ok(Self::Chr),
      "cos" => Ok(Self::Cos),
      "cvi$" => Ok(Self::Cvi),
      "cvs$" => Ok(Self::Cvs),
      "eof" => Ok(Self::Eof),
      "exp" => Ok(Self::Exp),
      "int" => Ok(Self::Int),
      "left$" => Ok(Self::Left),
      "len" => Ok(Self::Len),
      "lof" => Ok(Self::Lof),
      "log" => Ok(Self::Log),
      "mid$" => Ok(Self::Mid),
      "mki$" => Ok(Self::Mki),
      "mks$" => Ok(Self::Mks),
      "peek" => Ok(Self::Peek),
      "pos" => Ok(Self::Pos),
      "right$" => Ok(Self::Right),
      "rnd" => Ok(Self::Rnd),
      "sgn" => Ok(Self::Sgn),
      "sin" => Ok(Self::Sin),
      "sqr" => Ok(Self::Sqr),
      "str$" => Ok(Self::Str),
      "tan" => Ok(Self::Tan),
      "val" => Ok(Self::Val),
      _ => Err(()),
    }
  }
}

impl Debug for BinaryOpKind {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    let kind = match self {
      Self::Eq => "=",
      Self::Ne => "<>",
      Self::Gt => ">",
      Self::Lt => "<",
      Self::Ge => ">=",
      Self::Le => "<=",
      Self::Add => "+",
      Self::Sub => "-",
      Self::Mul => "*",
      Self::Div => "/",
      Self::Pow => "^",
      Self::And => "AND",
      Self::Or => "OR",
    };
    write!(f, "{}", kind)
  }
}

impl Debug for SysFuncKind {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    let kind = match self {
      Self::Abs => "ABS",
      Self::Asc => "ASC",
      Self::Atn => "ATN",
      Self::Chr => "CHR$",
      Self::Cos => "COS",
      Self::Cvi => "CVI$",
      Self::Cvs => "CVS$",
      Self::Eof => "EOF",
      Self::Exp => "EXP",
      Self::Int => "INT",
      Self::Left => "LEFT$",
      Self::Len => "LEN",
      Self::Lof => "LOF",
      Self::Log => "LOG",
      Self::Mid => "MID$",
      Self::Mki => "MKI$",
      Self::Mks => "MKS$",
      Self::Peek => "PEEK",
      Self::Pos => "POS",
      Self::Right => "RIGHT$",
      Self::Rnd => "RND",
      Self::Sgn => "SGN",
      Self::Sin => "SIN",
      Self::Sqr => "SQR",
      Self::Str => "STR$",
      Self::Tan => "TAN",
      Self::Val => "VAL",
    };
    write!(f, "{}", kind)
  }
}

impl Debug for UnaryOpKind {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    let kind = match self {
      Self::Neg => "-",
      Self::Not => "NOT",
      Self::Pos => "+",
    };
    write!(f, "{}", kind)
  }
}

impl Expr {
  pub fn print(
    &self,
    expr_arena: &Arena<Expr>,
    text: &str,
    f: &mut impl Write,
  ) -> fmt::Result {
    let range = self.range.clone();
    match &self.kind {
      ExprKind::Ident => write!(f, "<ID: {}>", &text[range.start..range.end]),
      ExprKind::StringLit => {
        write!(f, "<STR: {}>", &text[range.start..range.end])
      }
      ExprKind::NumberLit => {
        write!(f, "<NUM: {}>", &text[range.start..range.end])
      }
      ExprKind::SysFuncCall {
        func: (func_range, kind),
        args,
      } => {
        assert_eq!(
          text[func_range.start..func_range.end].to_ascii_uppercase(),
          format!("{:?}", kind)
        );
        write!(f, "{:?}(", kind)?;
        let mut comma = false;
        for &arg in args.iter() {
          if comma {
            write!(f, ", ")?;
          }
          comma = true;
          expr_arena[arg].print(expr_arena, text, f)?;
        }
        write!(f, ")")
      }
      ExprKind::UserFuncCall { func, arg } => {
        if let Some(func) = func {
          write!(f, "FN {}(", &text[func.start..func.end])?;
        } else {
          write!(f, "FN ???(")?;
        }
        expr_arena[*arg].print(expr_arena, text, f)?;
        write!(f, ")")
      }
      ExprKind::Binary {
        lhs,
        op: (op_range, kind),
        rhs,
      } => {
        assert_eq!(
          text[op_range.start..op_range.end]
            .to_owned()
            .replace(" ", "")
            .to_ascii_uppercase(),
          format!("{:?}", kind)
        );
        write!(f, "(")?;
        expr_arena[*lhs].print(expr_arena, text, f)?;
        write!(f, " {:?} ", kind)?;
        expr_arena[*rhs].print(expr_arena, text, f)?;
        write!(f, ")")
      }
      ExprKind::Unary {
        op: (op_range, kind),
        arg,
      } => {
        assert_eq!(
          text[op_range.start..op_range.end]
            .to_owned()
            .to_ascii_uppercase(),
          format!("{:?}", kind)
        );
        write!(f, "({:?} ", kind)?;
        expr_arena[*arg].print(expr_arena, text, f)?;
        write!(f, ")")
      }
      ExprKind::Index { name, indices } => {
        if let Some(name) = name {
          write!(f, "{}[", &text[name.start..name.end])?;
        } else {
          write!(f, "???[")?;
        }
        let mut comma = false;
        for &arg in indices.iter() {
          if comma {
            write!(f, ", ")?;
          }
          comma = true;
          expr_arena[arg].print(expr_arena, text, f)?;
        }
        write!(f, "]")
      }
      ExprKind::Inkey => write!(f, "<INKEY$>"),
      ExprKind::Error => write!(f, "<ERROR>"),
    }
  }
}
