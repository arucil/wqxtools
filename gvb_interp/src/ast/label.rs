use std::{num::IntErrorKind, str::FromStr};

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Label(pub u16);

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ParseLabelError {
  OutOfBound,
  NotALabel,
}

impl FromStr for Label {
  type Err = ParseLabelError;

  fn from_str(s: &str) -> Result<Self, Self::Err> {
    match s.parse::<u16>() {
      Ok(label) => {
        if label > 9999 {
          Err(ParseLabelError::OutOfBound)
        } else {
          Ok(Self(label))
        }
      }
      Err(err) => match err.kind() {
        IntErrorKind::PosOverflow => Err(ParseLabelError::OutOfBound),
        _ => Err(ParseLabelError::NotALabel),
      },
    }
  }
}
