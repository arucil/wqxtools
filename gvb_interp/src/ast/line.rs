#![allow(clippy::write_with_newline)]

use super::{Label, Range, StmtId};
use smallvec::SmallVec;
#[cfg(test)]
use std::fmt::Write;
use std::fmt::{self, Debug, Display, Formatter};

#[cfg(test)]
use widestring::Utf16Str;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ProgramLine {
  /// Includes newline.
  pub source_len: usize,
  pub label: Option<(Range, Label)>,
  pub stmts: SmallVec<[StmtId; 1]>,
  pub eol: Eol,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Eol {
  /// Only applies to last line.
  None,
  Lf,
  CrLf,
}

#[cfg(test)]
impl crate::parser::ParseResult<ProgramLine> {
  pub fn to_string(&self, text: &Utf16Str) -> String {
    let mut f = String::new();
    writeln!(&mut f, "label: {:?}", self.content.label).unwrap();
    writeln!(&mut f, "len: {}", self.content.source_len).unwrap();
    writeln!(&mut f, "eol: {:?}", self.content.eol).unwrap();
    writeln!(&mut f, "diagnostics: ").unwrap();
    for diag in &self.diagnostics {
      writeln!(&mut f, "  {:?}", diag).unwrap();
    }
    writeln!(&mut f, "-----------------").unwrap();
    for &stmt in self.content.stmts.iter() {
      self.stmt_arena[stmt]
        .print(&self.stmt_arena, &self.expr_arena, text, &mut f)
        .unwrap();
    }
    f
  }
}

impl Display for Eol {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    match self {
      Self::None => Ok(()),
      Self::Lf => write!(f, "\n"),
      Self::CrLf => write!(f, "\r\n"),
    }
  }
}

impl Eol {
  pub fn byte_len(&self) -> usize {
    match self {
      Self::None => 0,
      Self::Lf => 1,
      Self::CrLf => 2,
    }
  }
}
