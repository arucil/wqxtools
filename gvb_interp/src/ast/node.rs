use super::Range;
use id_arena::Id;

pub type NodeId = Id<Node>;

#[derive(Debug, Clone)]
pub struct Node {
  pub kind: NodeKind,
  pub range: Range,
  pub is_recovered: bool,
}

#[derive(Debug, Clone)]
pub enum NodeKind {
  Expr(super::expr::Expr),
  Stmt(super::stmt::Stmt),
}
