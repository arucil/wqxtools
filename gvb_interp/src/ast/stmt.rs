use super::{ExprId, Label, NonEmptyVec, Range, StmtId};
use smallvec::SmallVec;

#[derive(Debug, Clone)]
pub struct Stmt {
  pub kind: StmtKind,
  pub range: Range,
  pub is_recovered: bool,
}

#[derive(Debug, Clone)]
pub enum StmtKind {
  /// identical to REM
  Auto(Range),
  Beep,
  Box(NonEmptyVec<[ExprId; 4]>),
  Call(ExprId),
  Circle(NonEmptyVec<[ExprId; 3]>),
  Clear,
  Close {
    filenum: ExprId,
  },
  Cls,
  Cont,
  /// identical to REM
  Copy(Range),
  Data(NonEmptyVec<[Datum; 1]>),
  Def {
    /// ident
    name: Range,
    /// ident
    param: Range,
    body: ExprId,
  },
  /// identical to REM
  Del(Range),
  /// lvalue list
  Dim(NonEmptyVec<[ExprId; 1]>),
  Draw(NonEmptyVec<[ExprId; 2]>),
  /// identical to REM
  Edit(Range),
  Ellipse(NonEmptyVec<[ExprId; 4]>),
  End,
  Field {
    filenum: ExprId,
    fields: NonEmptyVec<[FieldSpec; 1]>,
  },
  /// identical to REM
  Files(Range),
  Flash,
  For {
    /// ident
    var: Range,
    start: ExprId,
    end: ExprId,
    step: Option<ExprId>,
  },
  Get {
    filenum: ExprId,
    record: ExprId,
  },
  GoSub(Option<(Range, Label)>),
  GoTo {
    has_goto_keyword: bool,
    label: Option<(Range, Label)>,
  },
  Graph,
  If {
    cond: ExprId,
    conseq: SmallVec<[StmtId; 1]>,
    alt: Option<SmallVec<[StmtId; 1]>>,
  },
  InKey,
  Input {
    source: InputSource,
    /// lvalue list
    vars: NonEmptyVec<[ExprId; 1]>,
  },
  Inverse,
  /// identical to REM
  Kill(Range),
  Let {
    /// lvalue
    var: ExprId,
    value: ExprId,
  },
  Line(NonEmptyVec<[ExprId; 4]>),
  /// identical to REM
  List(Range),
  /// identical to REM
  Load(Range),
  Locate {
    row: Option<ExprId>,
    column: Option<ExprId>,
  },
  LSet {
    /// lvalue
    var: ExprId,
    value: ExprId,
  },
  /// identical to REM
  New(Range),
  Next {
    /// ident list. may be empty
    vars: SmallVec<[Range; 1]>,
  },
  Normal,
  NoTrace,
  On {
    cond: ExprId,
    labels: NonEmptyVec<[(Range, Label); 2]>,
    is_sub: bool,
  },
  Open {
    filename: ExprId,
    mode: FileMode,
    filenum: ExprId,
    len: Option<ExprId>,
  },
  Play(ExprId),
  Poke {
    addr: ExprId,
    value: ExprId,
  },
  Pop,
  Print(SmallVec<[PrintElement; 2]>),
  Put {
    filenum: ExprId,
    record: ExprId,
  },
  /// lvalue list
  Read(NonEmptyVec<[ExprId; 1]>),
  Rem(Range),
  /// identical to REM
  Rename(Range),
  Restore(Option<(Range, Label)>),
  Return,
  RSet {
    /// lvalue
    var: ExprId,
    value: ExprId,
  },
  Run,
  /// identical to REM
  Save(Range),
  /// identical to REM
  Stop(Range),
  Swap {
    left: ExprId,
    right: ExprId,
  },
  System,
  Text,
  Trace,
  Wend,
  While(ExprId),
  Write {
    filenum: Option<ExprId>,
    data: NonEmptyVec<[WriteElement; 1]>,
  },
  NoOp,
}

#[derive(Debug, Clone)]
pub struct Datum {
  /// Does not include quotes.
  pub range: Range,
  pub is_quoted: bool,
}

#[derive(Debug, Clone)]
pub struct FieldSpec {
  pub range: Range,
  pub len: ExprId,
  /// lvalue
  pub var: ExprId,
}

#[derive(Debug, Clone)]
pub enum InputSource {
  /// file num expr
  File(ExprId),
  /// prompt string literal
  Keyboard(ExprId),
  Error,
}

#[derive(Debug, Clone)]
pub enum FileMode {
  Input,
  Output,
  Append,
  Random,
  Error,
}

#[derive(Debug, Clone)]
pub enum PrintElement {
  Expr(ExprId),
  Comma,
  Semicolon,
}

#[derive(Debug, Clone)]
pub struct WriteElement {
  pub datum: ExprId,
  pub comma: bool,
}