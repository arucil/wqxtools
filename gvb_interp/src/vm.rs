use std::convert::TryFrom;
use std::num::NonZeroUsize;

use crate::ast::Range;
use crate::util::mbf5::FloatError;
use crate::util::mbf5::Mbf5;
use crate::util::mbf5::Mbf5Accum;
use crate::HashMap;

pub(crate) use self::codegen::*;
pub(crate) use self::file::*;
pub(crate) use self::instruction::*;
pub(crate) use self::memory::*;
pub(crate) use self::r#type::*;

pub(crate) mod codegen;
pub(crate) mod file;
pub(crate) mod instruction;
pub(crate) mod memory;
pub(crate) mod r#type;

use string_interner::DefaultSymbol as Symbol;
use string_interner::StringInterner;

#[derive(Debug, Clone)]
pub(crate) struct Datum {
  pub range: Range,
  /// Does not include quotes.
  pub value: ByteString,
  pub is_quoted: bool,
}

pub struct VirtualMachine {
  data: Vec<Datum>,
  data_ptr: usize,
  pc: usize,
  code: Vec<Instr>,
  screen_mode: ScreenMode,
  print_mode: PrintMode,
  control_stack: Vec<ControlStackItem>,
  value_stack: Vec<(Range, TmpValue)>,
  interner: StringInterner,
  vars: HashMap<Symbol, Value>,
  arrays: HashMap<Symbol, Array>,
  user_funcs: HashMap<Symbol, UserFunc>,
  memory_man: MemoryManager,
  file_man: FileManager,
  state: ExecState,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum Type {
  Integer,
  Real,
  String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum ExecState {
  Done,
  Normal,
  WaitForKeyboardInput,
  AsmSuspend(),
}

#[derive(Debug, Clone)]
enum ControlStackItem {
  ForLoop {
    addr: Addr,
    var: Symbol,
    target: Mbf5Accum,
    step: Mbf5Accum,
  },
  WhileLoop {
    addr: Addr,
  },
  Sub {
    next_addr: Addr,
  },
}

#[derive(Debug, Clone)]
enum TmpValue {
  LValue(LValue),
  String(ByteString),
  Real(Mbf5Accum),
}

#[derive(Debug, Clone)]
enum LValue {
  Index { name: Symbol, offset: usize },
  Var { name: Symbol },
  Fn { name: Symbol, param: Symbol },
}

/// persistent value
#[derive(Debug, Clone)]
pub enum Value {
  Integer(u16),
  Real(Mbf5),
  String(ByteString),
}

#[derive(Debug, Clone)]
struct Array {
  bounds: Vec<NonZeroUsize>,
  data: ArrayData,
}

#[derive(Debug, Clone)]
enum ArrayData {
  Integer(Vec<u16>),
  Real(Vec<Mbf5>),
  String(Vec<ByteString>),
}

struct UserFunc {
  param: Symbol,
  body_addr: Addr,
}

#[derive(Debug, Clone)]
pub enum ExecResult {
  End,
  Continue,
  Sleep(usize),
  KeyboardInput {
    prompt: Option<String>,
    fields: Vec<KeyboardInputType>,
  },
  Error {
    range: Range,
    message: String,
  },
}

#[derive(Debug, Clone)]
pub enum KeyboardInputType {
  String,
  Real,
  Func { name: String, param: String },
}

#[derive(Debug, Clone)]
pub enum ExecInput {
  KeyboardInput(Vec<KeyboardInput>),
}

#[derive(Debug, Clone)]
pub enum KeyboardInput {
  String(ByteString),
  Real(Mbf5),
  Func {
    name: String,
    param: String,
    body: (),
  },
}

impl VirtualMachine {
  pub fn new(
    g: CodeGen,
    memory_man: MemoryManager,
    file_man: FileManager,
  ) -> Self {
    Self {
      data: g.data,
      data_ptr: 0,
      pc: 0,
      code: g.code,
      screen_mode: ScreenMode::Text,
      print_mode: PrintMode::Normal,
      control_stack: vec![],
      value_stack: vec![],
      interner: g.interner,
      vars: HashMap::default(),
      arrays: HashMap::default(),
      user_funcs: HashMap::default(),
      memory_man,
      file_man,
      state: ExecState::Normal,
    }
  }

  pub fn exec(&mut self, input: Option<ExecInput>, steps: usize) -> ExecResult {
    match self.state {
      ExecState::Done => return ExecResult::End,
      ExecState::WaitForKeyboardInput => {
        todo!()
      }
      ExecState::AsmSuspend() => {
        todo!()
      }
      ExecState::Normal => {
        // do nothing
      }
    }
    for _ in 0..steps {
      if let Err(result) = self.exec_instr() {
        return result;
      }
    }
    ExecResult::Continue
  }

  fn exec_instr(&mut self) -> Result<(), ExecResult> {
    let instr = &self.code[self.pc];
    let range = instr.range.clone();
    match instr.kind.clone() {
      InstrKind::DefFn { name, param, end } => {
        self.user_funcs.insert(
          name,
          UserFunc {
            param,
            body_addr: Addr(self.pc + 1),
          },
        );
        self.pc = end.0;
        return Ok(());
      }
      InstrKind::DimArray { name, dimensions } => {
        if self.arrays.contains_key(&name) {
          self.state.error(range, "重复定义数组")?;
        }
        let mut size = 1;
        let mut bounds = vec![];
        for _ in 0..dimensions.get() {
          let (range, value) = self.value_stack.pop().unwrap();
          let value = value.unwrap_real();
          let bound = f64::from(value.truncate()) as isize;
          if bound < 0 {
            self.state.error(
              range,
              format!("数组下标不能为负数。该下标的值为：{}", f64::from(value)),
            )?
          }
          size *= bound as usize;
          bounds
            .push(unsafe { NonZeroUsize::new_unchecked(bound as usize + 1) });
        }
        let data = ArrayData::new(symbol_type(&self.interner, name), size);
        self.arrays.insert(name, Array { bounds, data });
      }
      InstrKind::PushLValue { name, dimensions } => {
        if dimensions == 0 {
          self
            .value_stack
            .push((range, TmpValue::LValue(LValue::Var { name })));
        } else {
          let offset = self.calc_array_offset(name, dimensions)?;
          self
            .value_stack
            .push((range, TmpValue::LValue(LValue::Index { name, offset })));
        }
      }
      InstrKind::PushFnLValue { name, param } => {
        self
          .value_stack
          .push((range, TmpValue::LValue(LValue::Fn { name, param })));
      }
      InstrKind::SetRecordFields { .. } => todo!(),
      InstrKind::ForLoop { name, has_step } => {
        let step = if has_step {
          self.value_stack.pop().unwrap().1.unwrap_real()
        } else {
          Mbf5Accum::one()
        };
        let end = self.value_stack.pop().unwrap().1.unwrap_real();
        let start = self.value_stack.pop().unwrap().1.unwrap_real();

        let mut prev_loop = None;
        for (i, item) in self.control_stack.iter().enumerate().rev() {
          if let ControlStackItem::ForLoop { var: prev_var, .. } = item {
            if name == *prev_var {
              prev_loop = Some(i);
              break;
            }
          }
        }
        if let Some(i) = prev_loop {
          self.control_stack.truncate(i);
        }

        self.control_stack.push(ControlStackItem::ForLoop {
          addr: Addr(self.pc),
          var: name,
          target: end,
          step,
        });

        //self.store_lvalue(LValue::Var { name }, LValue::);
      }
      InstrKind::Assign => {
        let (rvalue_range, rvalue) = self.value_stack.pop().unwrap();
        let (lvalue_range, lvalue) = self.value_stack.pop().unwrap();
        let lvalue = lvalue.unwrap_lvalue();
        let rvalue = Value::from(rvalue);
        self.store_lvalue(lvalue, rvalue)?;
      }
      _ => todo!(),
    }
    self.pc += 1;
    Ok(())
  }

  fn calc_array_offset(
    &mut self,
    name: Symbol,
    dimensions: usize,
  ) -> Result<usize, ExecResult> {
    if !self.arrays.contains_key(&name) {
      let data = ArrayData::new(
        symbol_type(&self.interner, name),
        11usize.pow(dimensions as u32),
      );
      self.arrays.insert(
        name,
        Array {
          bounds: vec![unsafe { NonZeroUsize::new_unchecked(11) }; dimensions],
          data,
        },
      );
    }

    let array = &self.arrays[&name];
    let mut offset = 0;
    for i in (0..dimensions).rev() {
      let (range, value) = self.value_stack.pop().unwrap();
      let value = value.unwrap_real();
      let bound = f64::from(value.truncate()) as isize;
      if bound < 0 {
        self.state.error(
          range,
          format!(
            "数组下标不能为负数。该下标的值为：{}，取整后的值为：{}",
            f64::from(value),
            bound
          ),
        )?
      } else if bound as usize >= array.bounds[i].get() {
        self.state.error(
          range,
          format!(
            "数组下标超出上限。该下标的上限为：{}，该下标的值为：{}, 取整后的值为：{}",
            array.bounds[i].get(),
            f64::from(value),
            bound
          ),
        )?
      }

      offset = offset * array.bounds.get(i + 1).map_or(1, |n| n.get())
        + bound as usize;
    }
    Ok(offset)
  }

  fn store_lvalue(
    &mut self,
    lvalue: LValue,
    (rvalue_range, rvalue): (Range, TmpValue),
  ) -> Result<(), ExecResult> {
    macro_rules! assign_real {
      ($var:ident => $body:expr) => {
        let value = rvalue.unwrap_real();
        match Mbf5::try_from(value) {
          Ok($var) => {
            $body
          }
          Err(FloatError::Infinite) => {
            self.state.error(
              rvalue_range,
              format!(
                "运算结果数值过大，超出了实数的表示范围。运算结果为：{}",
                f64::from(value),
              ),
            )?;
          }
          Err(_) => unreachable!(),
        }
      }
    }
    match lvalue {
      LValue::Var { name } => {
        match symbol_type(&self.interner, name) {
          Type::Integer => {
            let value = rvalue.unwrap_real();
            if let Err(FloatError::Infinite) = Mbf5::try_from(value) {
              self.state.error(
                rvalue_range,
                format!(
                  "运算结果数值过大，超出了实数的表示范围。运算结果为：{}",
                  f64::from(value),
                ),
              )?;
            }
            let int = f64::from(value.truncate());
            if int <= -32769.0 || int >= 32768.0 {
              self.state.error(
                rvalue_range,
                format!(
                  "运算结果数值过大，超出了整数的表示范围（-32768~32767），\
                  无法赋值给整数变量。运算结果为：{}",
                  f64::from(value),
                ),
              )?;
            }
            self.vars.insert(name, Value::Integer(int as u16));
          }
          Type::Real => {
            assign_real!(num => self.vars.insert(name, Value::Real(num)));
          }
          Type::String => {
            self
              .vars
              .insert(name, Value::String(rvalue.unwrap_string()));
          }
        }
        Ok(())
      }
      LValue::Index { name, offset } => {
        match &mut self.arrays.get_mut(&name).unwrap().data {
          ArrayData::Integer(arr) => {
          }
          (ArrayData::Real(arr), Value::Real(n)) => arr[offset] = n,
          (ArrayData::String(arr), Value::String(n)) => arr[offset] = n,
          _ => unreachable!(),
        }
        Ok(())
      }
      LValue::Fn { .. } => unreachable!(),
    }
  }
}

impl ExecState {
  fn error<S: ToString>(
    &mut self,
    range: Range,
    message: S,
  ) -> Result<!, ExecResult> {
    *self = Self::Done;
    Err(ExecResult::Error {
      range,
      message: message.to_string(),
    })
  }
}

impl TmpValue {
  fn unwrap_real(self) -> Mbf5Accum {
    match self {
      Self::Real(num) => num,
      _ => unreachable!(),
    }
  }

  fn unwrap_string(self) -> ByteString {
    match self {
      Self::String(s) => s,
      _ => unreachable!(),
    }
  }

  fn unwrap_lvalue(self) -> LValue {
    match self {
      Self::LValue(lval) => lval,
      _ => unreachable!(),
    }
  }
}

fn symbol_type(interner: &StringInterner, symbol: Symbol) -> Type {
  match interner.resolve(symbol).unwrap().as_bytes().last().unwrap() {
    b'%' => Type::Integer,
    b'$' => Type::String,
    _ => Type::Real,
  }
}

impl ArrayData {
  fn new(ty: Type, size: usize) -> Self {
    match ty {
      Type::Integer => ArrayData::Integer(vec![0; size]),
      Type::Real => ArrayData::Real(vec![Mbf5::zero(); size]),
      Type::String => ArrayData::String(vec![ByteString::new(); size]),
    }
  }
}
