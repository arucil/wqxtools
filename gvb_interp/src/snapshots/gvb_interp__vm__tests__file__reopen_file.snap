---
source: gvb_interp/src/vm.rs
expression: device.log.borrow()

---
open file "f.DAT", read: false, write: true, truncate: true
close file

